mainnet_contract/contracts/LiteCheckpoint.sol

```diff
@@ -0,0 +1,424 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.19;
+
+import {HeaderReader} from "./libraries/HeaderReader.sol";
+
+contract LiteCheckpoint {
+    struct HeaderInfo {
+        uint64 number;
+        uint64 roundNum;
+        int64 mainnetNum;
+    }
+
+    struct UnCommittedHeaderInfo {
+        uint64 sequence;
+        uint64 lastRoundNum;
+        uint64 lastNum;
+    }
+
+    struct BlockLite {
+        bytes32 blockHash;
+        uint64 number;
+    }
+
+    struct Validators {
+        address[] set;
+        int256 threshold;
+    }
+    mapping(bytes32 => bytes32) private unCommittedLastHash;
+    mapping(bytes32 => uint256) private unCommittedTree; // padding uint64 | uint64 sequence | uint64 lastRoundNum | uint64 lastNum
+    mapping(bytes32 => uint256) private headerTree; // padding uint64 | uint64 number | uint64 roundNum | int64 mainnetNum
+    mapping(uint64 => bytes32) private heightTree;
+    bytes32[] private currentTree;
+    mapping(address => bool) private lookup;
+    mapping(address => bool) private uniqueAddr;
+    mapping(int256 => Validators) private validators;
+    Validators private currentValidators;
+    bytes32 private latestEpoch;
+    bytes32 private latestFinalizedBlock;
+
+    string public constant MODE = "lite";
+    uint64 public epochNum;
+    uint64 public immutable INIT_GAP;
+    uint64 public immutable INIT_EPOCH;
+
+    // Event types
+    event SubnetEpochBlockAccepted(bytes32 blockHash, uint64 number);
+
+    constructor(
+        address[] memory initialValidatorSet,
+        bytes memory block1,
+        uint64 initGap,
+        uint64 initEpoch
+    ) {
+        require(initialValidatorSet.length > 0, "Validator Empty");
+        bytes32 block1HeaderHash = keccak256(block1);
+        validators[1] = Validators({
+            set: initialValidatorSet,
+            threshold: int256((initialValidatorSet.length * 2 * 100) / 3)
+        });
+        currentValidators = validators[1];
+        setLookup(initialValidatorSet);
+        latestEpoch = block1HeaderHash;
+        INIT_GAP = initGap;
+        INIT_EPOCH = initEpoch;
+    }
+
+    /*
+     * @description core function in the contract, it can be summarized into three steps:
+     * 1. Verify subnet header meta information
+     * 2. Verify subnet header certificates
+     * 3. (Conditional) Update Committed Status for ancestor blocks
+     * @param list of rlp-encoded block headers.
+     */
+    function receiveHeader(bytes[] calldata headers) external {
+        require(
+            headers.length > 0,
+            "receiveHeader : Headers length must be greater than 0"
+        );
+        bytes memory header0 = headers[0];
+        saveEpoch(header0);
+        //for commit INIT_EPOCH
```

Comment:
> I don't understand this `if`

because the index 0 always is gap/epoch

for example someone only submmit 451 452 

451 will saveEpoch

452 will be sliceBytes function putin new array and exec commitHeader function